### PR TITLE
fix: enforce dead-code detection in CI and correct al doctor diagnostics

### PR DESCRIPTION
**Pull request by nicholasjconn**
## Summary

Autonomous `/improve-codebase` iteration #2, focused on dispatch / doctor / install / CI integrity. Each change fixes a silent failure mode or removes dead surface; no user-facing behavior changes beyond more accurate `al doctor` output and a CI gate that now actually fails on dead code.

## Highlights

1. **Dead-code check was a silent no-op.** The `deadcode` binary always exits 0, so `make dead-code` never failed regardless of findings. The target now captures `deadcode` output and fails on any non-empty result, and `dead-code` is wired into `make ci` so dead code blocks CI.
2. **`al doctor` malformed-`.env` now FAILs loudly.** A malformed/unreadable `.agent-layer/.env` previously got swallowed and surfaced as a misleading "Missing secret" cascade pointing at the wrong root cause; it now emits a Secrets FAIL with an actionable recommendation. A genuinely missing `.env` is still treated as valid (secrets may live in the process environment). Redundant, contradictory Secrets/Skills diagnostics are suppressed when `CheckConfig`'s lenient fallback already emitted a blocking FAIL, and the context size summary names skills as excluded (rather than silently counting 0 tokens) when skills fail to load.
3. **agentdispatch honors the dispatch exit-code contract.** stdout-write errors in the Claude and Codex stream decoders now carry `ExitTargetFailure` instead of a raw error that cobra would surface as a generic failure.
4. **Removed test-only dead code.** Deleted the unused `writeTemplateDir` function and the always-nil `recordDiff` parameter from `writeTemplateFile`, plus the tests that only exercised the dead paths.
5. **Message-constant additions.** `DispatchStdoutWriteFailedFmt`, `DoctorEnvFileUnreadableFmt`, `DoctorEnvFileUnreadableRecommend`, `DoctorSizeSkillsUnavailable`.
6. **Removed the stale `.golangci.yml` exclusion** for the deleted `writeTemplateDir` symbol.

## Deferred (ISSUES.md)

- **Dispatch lock-wait vs download-timeout tuning.** `withFileLock`'s lock-wait timeout currently equals the download timeout, so a waiter can fail while the holder is still legitimately downloading on a cold cache. Whether to bump the wait timeout is a contention-policy decision (waiters fail rather than pile up), left to the maintainer rather than guessed here.

## Summary by CodeRabbit

* **Bug Fixes**
  * Doctor command now reports configuration issues consistently when `.env` or skills fail to load, including accurate size-summary output for unavailable skills.
  * Improved dispatch stdout-write failures with clearer, target-specific error wrapping.

* **Tests**
  * Added doctor and config tests for lenient parsing and malformed `.env`/skills scenarios.
  * Tightened dispatch streaming error-chain assertions.

* **Documentation**
  * Documented a known dispatch lock-wait timeout edge case.

* **Chores**
  * CI dead-code checks now fail on findings; added to local `ci` target.
  * Updated lint configuration, and adjusted internal template-writing API with corresponding test updates/removals.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -116,14 +116,16 @@ test-race: ## Run race detector for concurrency-critical packages
 	@GOCACHE="$(GO_CACHE)" GOMODCACHE="$(GO_MOD_CACHE)" go test -race ./internal/sync/... ./internal/install/... ./internal/warnings/...
 
 .PHONY: dead-code
-dead-code: check-deadcode ## Run dead code analysis across all packages (test-aware)
+dead-code: check-deadcode ## Run dead code analysis across all packages (test-aware); fails on findings
 	@mkdir -p "$(GO_CACHE)" "$(GO_MOD_CACHE)"
-	@GOCACHE="$(GO_CACHE)" GOMODCACHE="$(GO_MOD_CACHE)" "$(TOOL_BIN)/deadcode" -test ./...
+	@out="$$(GOCACHE="$(GO_CACHE)" GOMODCACHE="$(GO_MOD_CACHE)" "$(TOOL_BIN)/deadcode" -test ./...)"; \
+	  if [[ -n "$$out" ]]; then echo "$$out" >&2; echo "dead code detected (deadcode always exits 0; non-empty output fails this target)" >&2; exit 1; fi
```
[the remark this answers is not in the document]
Comment (nicholasjconn) [reply Makefile]: **Fixed in `0bb1414`.** The target now captures combined stdout+stderr (`deadcode -test ./... 2>&1`) into `out` and records the command's exit status (`rc=$?`). It fails with `exit $rc` when deadcode itself exits non-zero (loader/compile error), and separately fails on non-empty `out` for the always-exit-0 findings case. A loader error on stderr now both lands in `out` and trips the explicit non-zero-exit check, so it can no longer silently pass the gate.

File: Makefile [hunk 2]
```diff
@@ -116,14 +116,16 @@ test-race: ## Run race detector for concurrency-critical packages
...
 
 .PHONY: dead-code-entrypoints
-dead-code-entrypoints: check-deadcode ## Run dead code analysis from CLI entrypoints only
+dead-code-entrypoints: check-deadcode ## Run dead code analysis from CLI entrypoints only; fails on findings
 	@mkdir -p "$(GO_CACHE)" "$(GO_MOD_CACHE)"
-	@GOCACHE="$(GO_CACHE)" GOMODCACHE="$(GO_MOD_CACHE)" "$(TOOL_BIN)/deadcode" -test ./cmd/al ./cmd/publish-site
+	@out="$$(GOCACHE="$(GO_CACHE)" GOMODCACHE="$(GO_MOD_CACHE)" "$(TOOL_BIN)/deadcode" -test ./cmd/al ./cmd/publish-site)"; \
+	  if [[ -n "$$out" ]]; then echo "$$out" >&2; echo "dead code detected (deadcode always exits 0; non-empty output fails this target)" >&2; exit 1; fi
```
[the remark this answers is not in the document]
Comment (nicholasjconn) [reply Makefile]: **Fixed in `0bb1414`.** Same hardening applied to `dead-code-entrypoints`: combined output capture via `2>&1` plus an explicit `rc=$?` check that fails with the deadcode exit code on a non-zero exit, before the non-empty-output findings check.

File: cmd/al/root_doctor_test.go [hunk 3]
```diff
@@ -563,6 +563,120 @@ func TestDoctorCommand_FlatSkillsDetectedEvenWhenConfigFails(t *testing.T) {
 	}
 }
 
+// writeTestRepoLenientConfig writes a repo whose config fails STRICT validation
+// (an unknown key) but still parses leniently, so doctor's lenient fallback runs
+// and the orchestrator's CheckSecrets/CheckSkills calls execute against the
+// partial config. body is appended verbatim under the valid sections.
```
[the remark this answers is not in the document]
Comment (nicholasjconn) [reply cmd/al/root_doctor_test.go]: **Fixed in `0bb1414`.** Removed the stale trailing sentence ("body is appended verbatim under the valid sections") from `writeTestRepoLenientConfig`'s doc comment; the helper takes only `(t, root)` and writes a fixed config, so the `body` reference was misleading.
